File: leetcode/easy/0543-diameter-of-binary-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int maxDiameter;
public:
    int getDepth(TreeNode *root)
    {
        if (root == nullptr)
        {
            return 0;
        }

        int leftPart = getDepth(root->left);
        int rightPart = getDepth(root->right);

        maxDiameter = max(maxDiameter, rightPart + leftPart);

        return max(leftPart ,rightPart) + 1;

    }
    int diameterOfBinaryTree(TreeNode *root)
    {
        maxDiameter = 0;
        getDepth(root);

        return maxDiameter;
    }
};

File: leetcode/easy/0543-diameter-of-binary-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    __maxDiameter = 0

    def get_depth(self, root: TreeNode) -> int:
        if root is None:
            return 0

        left_part = self.get_depth(root.left)
        right_part = self.get_depth(root.right)

        self.__maxDiameter = max(self.__maxDiameter, left_part + right_part)

        return max(left_part, right_part) + 1

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.__maxDiameter = 0

        self.get_depth(root)
        return self.__maxDiameter
